### PR TITLE
cpu/cortexm_common: split interrupt vectors into parts

### PR DESCRIPTION
**Pull request by sgso**
Proposal that splits the vectors on cortexm_common into a 'common' and a 'vendor specific' part which allows cpu implementations to just specify the vendor entries. 

- First commit contains the linker script changes with two sections (.vectors.common / .vectors.vendor) now instead of one and automatic initial stack pointer filling.

- Second commit changes the names of the stubs from *_fault_default  to *_fault and makes SVC, pendSV, NMI (renamed to isr_nmi) and SysTick calls to the shared dummy handler.

- Benefits:
 - Things are where they belong.
 - 20 lines of code less per cpu.
 - consistent names of ISRs enforced by code instead of conventions.
 - makes the ifdef'ing in platforms (sam3) less painful to watch.

- Third commit contains example for samd21.

- Might affect/ease  #3333 (why not just make it a module?)

- 4 Bytes overhead. I cut some strings for that.

### CODE REVIEW

File: cpu/cortexm_common/ldscripts/cortexm_base.ld [hunk 1]
```diff
@@ -88,6 +96,15 @@ SECTIONS
         _efixed = .;            /* End of text section */
     } > rom
 
+    /* Cortex-M expects the initial stack pointer at 0x0. */
+    ASSERT(_stack_base == ORIGIN(rom),
+           "ld script assert failed, check cortexm_base.ld for explanation.")
```
Comment (gebart) [inline cpu/cortexm_common/ldscripts/cortexm_base.ld]: could you make these assert messages a bit more specific?

File: cpu/cortexm_common/ldscripts/cortexm_base.ld [hunk 2]
```diff
@@ -88,6 +96,15 @@ SECTIONS
...
+    /* The vendor interrupt table has to start at 0x40. If it doesn't,
+       something is overflowing after the SysTick slot at 0x3c and this
+       assert will trigger. */
+    ASSERT(vectors_vendor == ORIGIN(rom) + 0x40,
+           "ld script assert failed, check cortexm_base.ld for explanation.")
```
Comment (gebart) [inline cpu/cortexm_common/ldscripts/cortexm_base.ld]: please move part of the explanation from the comment into the assert message

File: cpu/cortexm_common/ldscripts/cortexm_base.ld [hunk 3]
```diff
@@ -42,8 +42,16 @@ SECTIONS
     .text :
     {
         . = ALIGN(4);
-        _sfixed = .;
-        KEEP(*(.vectors .vectors.*))
+        _stack_base = .;
+        LONG(ABSOLUTE(_estack))
```
Comment (gebart) [inline cpu/cortexm_common/ldscripts/cortexm_base.ld]: while this is still a valid way of doing this I think that the readability of the code is better if the whole common part of the ISR vector table is inside `cpu/cortexm_common/vectors_cortexm.c` instead.

**Comment by gebart** [comment]
:+1: for this change in general. It is always good to consolidate things to avoid repeating ourselves.

**Comment by haukepetersen** [comment]
looks good to me

**Comment by sgso** [comment]
Adressed comments.

Can I start adapting the other cpus then?

**Comment by gebart** [comment]
@sgso yes, go ahead.

**Comment by sgso** [comment]
I took out the assertion completely because it screws up the output of objdump. Anybody knows how to fix it, so that objdump keeps showing the data as .word?

For most cpus it was straight-forward. For kinetis I only changed the handler names because it uses its own linker scripts. 

Somebody with a cc2538 or experience with it should test the new linker script. I don't think the separate linker script is necessary.

**Comment by haukepetersen** [comment]
>  155 additions and 538 deletions.

looking better and better :-)

**Comment by gebart** [comment]
@sgso I don't understand why the asserts would make objdump misbehave. What do you mean by screws up the output?

**Comment by sgso** [comment]
@gebart I don't know why this is happening:

`objdump -d` without assertion:

``` 
08000000 <_vectors_common>:
 8000000:       20000b68        .word   0x20000b68
 8000004:       08000465        .word   0x08000465
 8000008:       08000451        .word   0x08000451
 800000c:       08000441        .word   0x08000441
        ...
 800002c:       08000415        .word   0x08000415
        ...
 8000038:       080003f5        .word   0x080003f5
 800003c:       08000451        .word   0x08000451

08000040 <_vectors_vendor>:
 8000040:       0800015d        .word   0x0800015d
 8000044:       0800015d        .word   0x0800015d
```
Same output with `ASSERT(ABSOLUTE(_vectors_vendor) == ORIGIN(rom) + 0x40, "...")` added:
```
08000000 <_vectors_common>:                                                                    
 8000000:       20000b68        .word   0x20000b68                                             
 8000004:       08000465        .word   0x08000465                                             
 8000008:       08000451        .word   0x08000451                                             
 800000c:       08000441        .word   0x08000441                                             
        ...                                                                                    
 800002c:       08000415        .word   0x08000415                                             
        ...                                                                                    
 8000038:       080003f5        .word   0x080003f5                                             
 800003c:       08000451        .word   0x08000451                                            
                                                                                               
08000040 <interrupt_vector>:                                                             
 8000040:       0800015d 0800015d 0800015d 0800015d     ]...]...]...]...                       
 8000050:       0800015d 0800015d 0800015d 0800015d     ]...]...]...]...                       
 8000060:       0800015d 0800015d 0800015d 0800015d     ]...]...]...]...                       
 8000070:       0800015d 0800015d 0800015d 08001015     ]...]...].......                       
 8000080:       0800015d 0800015d 0800015d 0800015d     ]...]...]...]...                       
 8000090:       0800015d 0800015d 0800015d 0800015d     ]...]...]...]...                       
 80000a0:       0800015d 0800015d 0800015d 08000ebd     ]...]...].......                       
 80000b0:       08000ecd 0800015d 0800015d 00000000     ....]...].......  
```

**Comment by gebart** [comment]
I'm clueless as to the reason of the objdump difference.

However, after looking over the linker script again I think it might be better to put the vectors in their own section, like the kinetis scripts do it. It makes it easier to examine the binary with the usual tools, for example `objdump -s -j .vector_table`

**Comment by sgso** [comment]
I'm almost sure they will be scrambled like that as well when put in their own sections but there might be a way to mark them as .word-data.

I initially opted against this because it does not show up without extra arguments in objdump and might confuse people ("Who stole my vectors?"). I'd like a second opinion because I don't care that much, @haukepetersen?

**Comment by sgso** [comment]
Just tried, creating two sections .vector_table_common / .vector_table_vendor allows assertions on the size of the common part and does not scramble output.

**Comment by haukepetersen** [comment]
I am actually a fan of not putting the vectors in its own section, so it shows up in the default `make objdump`. But this is not a strong opinion - so whatever the majority votes for...

What I would love to see is a unification of linkerscripts between kinetis, cc2538 and the rest of the boards - as said, in either way... But this is not subject of this PR...

Regarding the scrambled object dump output: I have no idea...

**Comment by OlegHahm** [comment]
Please state a reason for closing.

**Pull request closed without merging**